### PR TITLE
fix: add readiness probe for lms/cms

### PR DESCRIPTION
**Pull request by Ian2012**
This PR:

- Disable uwsgi request logging
- Disable `tracking.log` and `all.log` files for lms and cms pods.
- Removes the prestop hook in favor of a uwsgi hook call to gracefully terminate workers, which improved the time needed to restart or terminate lms/cms pods while being able to answer pending requests
- Adds livenessProbe for pods.
- Adds a soft pod AntiAffinity to prefer scheduling pods across nodes.
- Setups maxUnavailable to 0 to safely rollout deployments without losing capacity.
- Setups maxSurge to 2 so as not to overcommit resources temporarily for a deployment.

### CODE REVIEW

File: drydock/templates/drydock/k8s/lifecycle/cms.yml [hunk 1]
```diff
@@ -7,21 +7,41 @@ spec:
...
+            podAffinityTerm:
+              labelSelector:
+                matchLabels:
+                  app.kubernetes.io/name: cms
+              topologyKey: kubernetes.io/hostname
+  strategy:
+    type: RollingUpdate
+    rollingUpdate:
+      maxUnavailable: 0
+      maxSurge: 2
```
Comment (MoisesGSalas) [inline drydock/templates/drydock/k8s/lifecycle/cms.yml:47]: Wouldn't this make the update really slow if we have, say, 100 pods or something like that? Shouldn't it be a percentage?

File: drydock/patches/uwsgi-config [hunk 2]
```diff
@@ -12,3 +12,5 @@ no-defer-accept = true
 master = true
 py-call-osafterfork = true
 vacuum = true
+hook-master-start = unix_signal:1 gracefully_kill_them_all
+disable-logging = true
```
Comment (MoisesGSalas) [inline drydock/patches/uwsgi-config:16]: How does the logs in stdout out look with all the settings applied?

Comment (Ian2012) [reply drydock/patches/uwsgi-config:16]: Only edx-platform stdout and stderr is available. It removes only request logs such as:

```
[pid: 9|app: 0|req: 2/3] 10.100.2.19 () {48 vars in 778 bytes} [Thu Sep  5 01:26:36 2024] GET / => generated 9508 bytes in 1955 msecs (HTTP/1.1 200) 7 headers in 663 bytes (1 switches on core 0)
```

File: drydock/templates/drydock/k8s/lifecycle/cms.yml (hunk 1 above)
Comment (Ian2012) [reply drydock/templates/drydock/k8s/lifecycle/cms.yml:47]: Yes, the default value is 25% which seems good for us

File: drydock/templates/drydock/k8s/lifecycle/lms.yml [hunk 3]
```diff
@@ -7,21 +7,41 @@ spec:
     spec:
       containers:
         - name: lms
-          lifecycle:
```
Comment (jfavellar90) [inline drydock/templates/drydock/k8s/lifecycle/lms.yml:10]: Do you have any reason to remove the lifecycle hook? I think we can reduce this hook to a wait of 15 seconds and set `terminationGracePeriodSeconds` to 30 (the default) to reduce the termination waiting time.

File: drydock/patches/openedx-common-settings [hunk 4]
```diff
@@ -18,3 +18,6 @@ XBLOCK_SETTINGS["ScormXBlock"] = {
     "STORAGE_FUNC": scorm_xblock_storage,
 }
 {% endif %}
+
+LOGGING.get("loggers", {}).pop("tracking")
```
Comment (MoisesGSalas) [inline drydock/patches/openedx-common-settings:22]: Why removing the tracking logger?. Also, do it using bracket indexing, if there's an error is better for it to fail early.

[the remark this answers is not in the document]
Comment (Ian2012) [reply drydock/patches/openedx-common-settings:22]: Tracking logs are still logged to the console, without it they were logged twice and emitted to a local log file which we don't use and increase pod disk usage.

File: drydock/templates/drydock/k8s/lifecycle/lms.yml (hunk 3 above)
Comment (Ian2012) [reply drydock/templates/drydock/k8s/lifecycle/lms.yml:10]: Based on our testing it seems this is not necessary for UWSGI deployments as the hook applied will first terminate gracefully the workers and then kill himself.

However, the should keep the `terminationGracePeriodSeconds`.

File: drydock/patches/openedx-common-settings (hunk 4 above)
Comment (Ian2012) [reply drydock/patches/openedx-common-settings:22]: This removes the duplicated logger and the `tracking.log` file.

File: drydock/templates/drydock/k8s/lifecycle/cms.yml [hunk 5]
```diff
@@ -12,16 +12,37 @@ spec:
               exec:
                 command:
                 - "sleep"
-                - "30"
+                - "15"
```
Comment (MoisesGSalas) [inline drydock/templates/drydock/k8s/lifecycle/cms.yml:15]: I still want to challenge this preStop hook. Why is that we need this? Why isn't the uWSGI finish routine not enough?

Comment (jfavellar90) [reply drydock/templates/drydock/k8s/lifecycle/cms.yml:15]: @Ian2012 was performing a few tests related to this feature, we should get good input from there soon.

File: drydock/templates/drydock/k8s/lifecycle/cms.yml [hunk 6]
```diff
@@ -28,7 +28,7 @@ spec:
             httpGet:
               httpHeaders:
                 - name: Host
-                  value: {{ LMS_HOST }}
+                  value: {{ CMS_HOST }}
```
Comment (MoisesGSalas) [inline drydock/templates/drydock/k8s/lifecycle/cms.yml:31]: what about using just `cms` (and `lms`)? They are already in the [`ALLOWED_HOSTS`](https://github.com/overhangio/tutor/blob/46b401690c812448c17b2cd91f90cf765a7a9678/tutor/templates/apps/openedx/settings/lms/production.py#L10) list and this way it reduces additional git diffs when doing changes.

File: drydock/templates/drydock/k8s/lifecycle/cms.yml (hunk 5 above)
Comment (jfavellar90) [reply drydock/templates/drydock/k8s/lifecycle/cms.yml:15]: After the tests, the behavior is slightly better. We decided to keep the `preStop` hook with a 15 seconds sleep period

File: drydock/templates/drydock/k8s/lifecycle/lms.yml [hunk 7]
```diff
@@ -7,21 +7,36 @@ spec:
...
```
Comment (jfavellar90) [inline drydock/templates/drydock/k8s/lifecycle/lms.yml:10]: @Ian2012 As we discussed, keep the lifecycle for LMS as well

**Review by jfavellar90 — APPROVED** [review]

**Review by MoisesGSalas — APPROVED** [review]

**Pull request merged**